java-extension/com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/util/TestSearchUtils.java

```diff
@@ -416,6 +425,131 @@ private static void findTestItemsInTypeBinding(ITypeBinding typeBinding, JavaTes
         return result;
     }
 
+    /**
+     * Given the test item's full name, get its location. This is used to calculate the location for each test message.
+     */
+    public static Location findTestLocation(List<Object> arguments, IProgressMonitor monitor)
+            throws JavaModelException {
+        final String fullName = (String) arguments.get(0);
+        if (StringUtils.isEmpty(fullName)) {
+            return null;
+        }
+
+        final int projectNameEnd = fullName.indexOf("@");
+        if (projectNameEnd < 0) {
+            return null;
+        }
+
+        final String projectName = fullName.substring(0, projectNameEnd);
+        if (StringUtils.isEmpty(projectName)) {
+            return null;
+        }
+
+        final IJavaProject javaProject = ProjectUtils.getJavaProject(projectName);
+        if (javaProject == null) {
+            return null;
+        }
+
+        final int methodStart = fullName.indexOf("#");
+        final String typeName;
+        final String methodName;
+        if (methodStart > 0) {
+            typeName = fullName.substring(projectNameEnd + 1, methodStart);
+            methodName = fullName.substring(methodStart + 1);
+        } else {
+            typeName = fullName.substring(projectNameEnd + 1);
+            methodName = null;
+        }
+
+        final IType type = findType(javaProject, typeName, monitor);
+        if (type == null) {
+            return null;
+        }
+
+        if (StringUtils.isEmpty(methodName)) {
+            return new Location(JDTUtils.getFileURI(type.getResource()), TestItemUtils.parseTestItemRange(type));
+        }
+
+        for (final IMethod method : type.getMethods()) {
+            if (methodName.equals(method.getElementName())) {
+                // TODO: handle the override method
+                return new Location(JDTUtils.getFileURI(method.getResource()),
+                        TestItemUtils.parseTestItemRange(method));
+            }
+        }
+
+        return null;
+    }
+
+    protected static final IType findType(final IJavaProject project, String className, IProgressMonitor monitor) {
+        final IType[] result = { null };
+        final String dottedName = className.replace('$', '.'); // for nested classes...
+        try {
+            if (project != null) {
+                result[0] = internalFindType(project, dottedName, new HashSet<IJavaProject>(), monitor);
+            }
+            if (result[0] == null) {
```

Comment:
this branch is never visisted, as `project` is not null when you call `findType`.

Reply:
it's because `internalFindType()` might return null. In that case, use the search engine as a fallback

Reply:
got it